\hypertarget{classHomologyMasker_1_1RefMasker}{\section{Homology\-Masker.\-Ref\-Masker Class Reference}
\label{classHomologyMasker_1_1RefMasker}\index{Homology\-Masker.\-Ref\-Masker@{Homology\-Masker.\-Ref\-Masker}}
}


Singleton class allowing to find D\-N\-A sequences homologies between a list of query and a subject and to write a modified version of the subject sequence.  




Inheritance diagram for Homology\-Masker.\-Ref\-Masker\-:
\nopagebreak
\begin{figure}[H]
\begin{center}
\leavevmode
\includegraphics[width=224pt]{classHomologyMasker_1_1RefMasker__inherit__graph}
\end{center}
\end{figure}


Collaboration diagram for Homology\-Masker.\-Ref\-Masker\-:
\nopagebreak
\begin{figure}[H]
\begin{center}
\leavevmode
\includegraphics[width=224pt]{classHomologyMasker_1_1RefMasker__coll__graph}
\end{center}
\end{figure}


\subsection{Detailed Description}
Singleton class allowing to find D\-N\-A sequences homologies between a list of query and a subject and to write a modified version of the subject sequence. 

First a blast database is created from a reference fasta file if it is not provided by the user. Then, query sequences fasta files are blasted against the newly created subject database. Finally, if blast hits are found (homologies) the subject genome is imported, hit locations are masked with \char`\"{}\-Ns\char`\"{} and a new masked reference is written 

Definition at line 27 of file Homology\-Masker.\-py.



The documentation for this class was generated from the following file\-:\begin{DoxyCompactItemize}
\item 
src/\hyperlink{HomologyMasker_8py}{Homology\-Masker.\-py}\end{DoxyCompactItemize}
